money-java-servlet/src/main/scala/com/comcast/money/java/servlet/TraceFilter.scala

```diff
@@ -16,52 +16,94 @@
 
 package com.comcast.money.java.servlet
 
-import javax.servlet._
-import javax.servlet.http.{ HttpServletRequest, HttpServletRequestWrapper, HttpServletResponse }
-
+import com.comcast.money.api.SpanId
 import com.comcast.money.core.Formatters._
 import com.comcast.money.core.Money
 import com.comcast.money.core.internal.SpanLocal
+import javax.servlet._
+import javax.servlet.http.{ HttpServletRequest, HttpServletRequestWrapper, HttpServletResponse }
 import org.slf4j.LoggerFactory
 
 import scala.util.{ Failure, Success }
 
 /**
  * A Java Servlet 2.5 Filter.  Examines the inbound http request, and will set the
- * trace context for the request if the money trace header is found
+ * trace context for the request if the money trace header or X-B3 style headers are found
  */
 class TraceFilter extends Filter {
 
   private val MoneyTraceHeader = "X-MoneyTrace"
+  private val B3TraceId = "X-B3-TraceId"
+  private val B3SpanId = "X-B3-SpanId"
+  private val B3ParentSpanId = "X-B3-ParentSpanId"
   private val logger = LoggerFactory.getLogger(classOf[TraceFilter])
   private val factory = Money.Environment.factory
 
   override def init(filterConfig: FilterConfig): Unit = {}
 
   override def destroy(): Unit = {}
 
+  private val spanName = "servlet"
+
   override def doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain): Unit = {
 
     SpanLocal.clear()
     val httpRequest = new HttpServletRequestWrapper(request.asInstanceOf[HttpServletRequest])
-    val incomingTraceId = Option(httpRequest.getHeader(MoneyTraceHeader)) map { incTrcaceId =>
-      // attempt to parse the incoming trace id (its a Try)
-      fromHttpHeader(incTrcaceId) match {
-        case Success(spanId) => SpanLocal.push(factory.newSpan(spanId, "servlet"))
-        case Failure(ex) => logger.warn("Unable to parse money trace for request header '{}'", incTrcaceId)
+
+    val spanId: Option[SpanId] = parseMoneyTraceHeader(httpRequest).orElse(parseB3TraceHeader(httpRequest))
+    spanId.foreach(s => SpanLocal.push(factory.newSpan(s, spanName)))
+
+    val httpResponse = response.asInstanceOf[HttpServletResponse]
+    setResponseHeader(MoneyTraceHeader)
+    setResponseHeader(B3TraceId)
+    setResponseHeader(B3ParentSpanId)
+    setResponseHeader(B3SpanId)
+
+    def setResponseHeader(headerName: String) = {
+      Option(httpRequest.getHeader(headerName)).foreach(v => httpResponse.addHeader(headerName, v))
+    }
+    chain.doFilter(request, response)
+  }
+
+  def parseMoneyTraceHeader(httpRequest: HttpServletRequest): Option[SpanId] = {
+
+    def parseHeader(headerValue: String): Option[SpanId] = {
+      fromHttpHeader(headerValue) match {
+        case Success(spanId) => Some(spanId)
+        case Failure(ex) => {
+          logger.warn("Unable to parse money trace for request header '{}'", headerValue)
+          None
+        }
```

Comment:
would personally prefer if the braces were dropped after the case

Reply:
I still feel a little bit that the parsers belong in Formatters but I am happy to defer to @paulcleary on that